File: Upnp/Device/Services/Cpp/Core/DvZappOrgTestWidgetController1.cpp
#include <Core/DvZappOrgTestWidgetController1.h>
#include <ZappTypes.h>
#include <Core/DvInvocationResponse.h>
#include <Service.h>
#include <FunctorDviInvocation.h>

using namespace Zapp;

DvProviderZappOrgTestWidgetController1::DvProviderZappOrgTestWidgetController1(DvDevice& aDevice)
    : DvProvider(aDevice.Device(), "zapp.org", "TestWidgetController", 1)
{
    
}

void DvProviderZappOrgTestWidgetController1::EnableActionCreateWidget()
{
    Zapp::Action* action = new Zapp::Action("CreateWidget");
    action->AddInputParameter(new ParameterString("WidgetUdn"));
    action->AddInputParameter(new ParameterUint("WidgetClass"));
    FunctorDviInvocation functor = MakeFunctorDviInvocation(*this, &DvProviderZappOrgTestWidgetController1::DoCreateWidget);
    iService->AddAction(action, functor);
}

void DvProviderZappOrgTestWidgetController1::EnableActionRemoveWidget()
{
    Zapp::Action* action = new Zapp::Action("RemoveWidget");
    action->AddInputParameter(new ParameterString("WidgetUdn"));
    FunctorDviInvocation functor = MakeFunctorDviInvocation(*this, &DvProviderZappOrgTestWidgetController1::DoRemoveWidget);
    iService->AddAction(action, functor);
}

void DvProviderZappOrgTestWidgetController1::EnableActionSetWidgetRegister()
{
    Zapp::Action* action = new Zapp::Action("SetWidgetRegister");
    action->AddInputParameter(new ParameterString("WidgetUdn"));
    action->AddInputParameter(new ParameterUint("RegisterIndex"));
    action->AddInputParameter(new ParameterUint("RegisterValue"));
    FunctorDviInvocation functor = MakeFunctorDviInvocation(*this, &DvProviderZappOrgTestWidgetController1::DoSetWidgetRegister);
    iService->AddAction(action, functor);
}

void DvProviderZappOrgTestWidgetController1::EnableActionGetWidgetRegister()
{
    Zapp::Action* action = new Zapp::Action("GetWidgetRegister");
    action->AddInputParameter(new ParameterString("WidgetUdn"));
    action->AddInputParameter(new ParameterUint("RegisterIndex"));
    action->AddOutputParameter(new ParameterUint("RegisterValue"));
    FunctorDviInvocation functor = MakeFunctorDviInvocation(*this, &DvProviderZappOrgTestWidgetController1::DoGetWidgetRegister);
    iService->AddAction(action, functor);
}

void DvProviderZappOrgTestWidgetController1::DoCreateWidget(IDviInvocation& aInvocation, TUint aVersion)
{
    aInvocation.InvocationReadStart();
    Brhz WidgetUdn;
    aInvocation.InvocationReadString("WidgetUdn", WidgetUdn);
    TUint WidgetClass = aInvocation.InvocationReadUint("WidgetClass");
    aInvocation.InvocationReadEnd();
    InvocationResponse resp(aInvocation);
    CreateWidget(resp, aVersion, WidgetUdn, WidgetClass);
}

void DvProviderZappOrgTestWidgetController1::DoRemoveWidget(IDviInvocation& aInvocation, TUint aVersion)
{
    aInvocation.InvocationReadStart();
    Brhz WidgetUdn;
    aInvocation.InvocationReadString("WidgetUdn", WidgetUdn);
    aInvocation.InvocationReadEnd();
    InvocationResponse resp(aInvocation);
    RemoveWidget(resp, aVersion, WidgetUdn);
}

void DvProviderZappOrgTestWidgetController1::DoSetWidgetRegister(IDviInvocation& aInvocation, TUint aVersion)
{
    aInvocation.InvocationReadStart();
    Brhz WidgetUdn;
    aInvocation.InvocationReadString("WidgetUdn", WidgetUdn);
    TUint RegisterIndex = aInvocation.InvocationReadUint("RegisterIndex");
    TUint RegisterValue = aInvocation.InvocationReadUint("RegisterValue");
    aInvocation.InvocationReadEnd();
    InvocationResponse resp(aInvocation);
    SetWidgetRegister(resp, aVersion, WidgetUdn, RegisterIndex, RegisterValue);
}

void DvProviderZappOrgTestWidgetController1::DoGetWidgetRegister(IDviInvocation& aInvocation, TUint aVersion)
{
    aInvocation.InvocationReadStart();
    Brhz WidgetUdn;
    aInvocation.InvocationReadString("WidgetUdn", WidgetUdn);
    TUint RegisterIndex = aInvocation.InvocationReadUint("RegisterIndex");
    aInvocation.InvocationReadEnd();
    InvocationResponse resp(aInvocation);
    InvocationResponseUint respRegisterValue(aInvocation, "RegisterValue");
    GetWidgetRegister(resp, aVersion, WidgetUdn, RegisterIndex, respRegisterValue);
}

void DvProviderZappOrgTestWidgetController1::CreateWidget(IInvocationResponse& /*aResponse*/, TUint /*aVersion*/, const Brx& /*aWidgetUdn*/, TUint /*aWidgetClass*/)
{
    ASSERTS();
}

void DvProviderZappOrgTestWidgetController1::RemoveWidget(IInvocationResponse& /*aResponse*/, TUint /*aVersion*/, const Brx& /*aWidgetUdn*/)
{
    ASSERTS();
}

void DvProviderZappOrgTestWidgetController1::SetWidgetRegister(IInvocationResponse& /*aResponse*/, TUint /*aVersion*/, const Brx& /*aWidgetUdn*/, TUint /*aRegisterIndex*/, TUint /*aRegisterValue*/)
{
    ASSERTS();
}

void DvProviderZappOrgTestWidgetController1::GetWidgetRegister(IInvocationResponse& /*aResponse*/, TUint /*aVersion*/, const Brx& /*aWidgetUdn*/, TUint /*aRegisterIndex*/, IInvocationResponseUint& /*aRegisterValue*/)
{
    ASSERTS();
}

File: Upnp/Device/Services/Cs/DvZappOrgTestWidgetController1.cs
using System;
using System.Runtime.InteropServices;
using System.Text;
using System.Collections.Generic;
using Zapp.Core;

namespace Zapp.Device.Providers
{
    public interface IDvProviderZappOrgTestWidgetController1 : IDisposable
    {
        
    }
    /// <summary>
    /// Provider for the zapp.org:TestWidgetController:1 UPnP service
    /// </summary>
    public class DvProviderZappOrgTestWidgetController1 : DvProvider, IDisposable, IDvProviderZappOrgTestWidgetController1
    {
        private GCHandle iGch;
        private ActionDelegate iDelegateCreateWidget;
        private ActionDelegate iDelegateRemoveWidget;
        private ActionDelegate iDelegateSetWidgetRegister;
        private ActionDelegate iDelegateGetWidgetRegister;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="aDevice">Device which owns this provider</param>
        protected DvProviderZappOrgTestWidgetController1(DvDevice aDevice)
            : base(aDevice, "zapp-org", "TestWidgetController", 1)
        {
            iGch = GCHandle.Alloc(this);
        }

        /// <summary>
        /// Signal that the action CreateWidget is supported.
        /// </summary>
        /// <remarks>The action's availability will be published in the device's service.xml.
        /// DoCreateWidget must be overridden if this is called.</remarks>
        protected unsafe void EnableActionCreateWidget()
        {
            Zapp.Core.Action action = new Zapp.Core.Action("CreateWidget");
            List<String> allowedValues = new List<String>();
            action.AddInputParameter(new ParameterString("WidgetUdn", allowedValues));
            action.AddInputParameter(new ParameterUint("WidgetClass"));
            iDelegateCreateWidget = new ActionDelegate(DoCreateWidget);
            EnableAction(action, iDelegateCreateWidget, GCHandle.ToIntPtr(iGch));
        }

        /// <summary>
        /// Signal that the action RemoveWidget is supported.
        /// </summary>
        /// <remarks>The action's availability will be published in the device's service.xml.
        /// DoRemoveWidget must be overridden if this is called.</remarks>
        protected unsafe void EnableActionRemoveWidget()
        {
            Zapp.Core.Action action = new Zapp.Core.Action("RemoveWidget");
            List<String> allowedValues = new List<String>();
            action.AddInputParameter(new ParameterString("WidgetUdn", allowedValues));
            iDelegateRemoveWidget = new ActionDelegate(DoRemoveWidget);
            EnableAction(action, iDelegateRemoveWidget, GCHandle.ToIntPtr(iGch));
        }

        /// <summary>
        /// Signal that the action SetWidgetRegister is supported.
        /// </summary>
        /// <remarks>The action's availability will be published in the device's service.xml.
        /// DoSetWidgetRegister must be overridden if this is called.</remarks>
        protected unsafe void EnableActionSetWidgetRegister()
        {
            Zapp.Core.Action action = new Zapp.Core.Action("SetWidgetRegister");
            List<String> allowedValues = new List<String>();
            action.AddInputParameter(new ParameterString("WidgetUdn", allowedValues));
            action.AddInputParameter(new ParameterUint("RegisterIndex"));
            action.AddInputParameter(new ParameterUint("RegisterValue"));
            iDelegateSetWidgetRegister = new ActionDelegate(DoSetWidgetRegister);
            EnableAction(action, iDelegateSetWidgetRegister, GCHandle.ToIntPtr(iGch));
        }

        /// <summary>
        /// Signal that the action GetWidgetRegister is supported.
        /// </summary>
        /// <remarks>The action's availability will be published in the device's service.xml.
        /// DoGetWidgetRegister must be overridden if this is called.</remarks>
        protected unsafe void EnableActionGetWidgetRegister()
        {
            Zapp.Core.Action action = new Zapp.Core.Action("GetWidgetRegister");
            List<String> allowedValues = new List<String>();
            action.AddInputParameter(new ParameterString("WidgetUdn", allowedValues));
            action.AddInputParameter(new ParameterUint("RegisterIndex"));
            action.AddOutputParameter(new ParameterUint("RegisterValue"));
            iDelegateGetWidgetRegister = new ActionDelegate(DoGetWidgetRegister);
            EnableAction(action, iDelegateGetWidgetRegister, GCHandle.ToIntPtr(iGch));
        }

        /// <summary>
        /// CreateWidget action.
        /// </summary>
        /// <remarks>Will be called when the device stack receives an invocation of the
        /// CreateWidget action for the owning device.
        ///
        /// Must be implemented iff EnableActionCreateWidget was called.</remarks>
        /// <param name="aVersion">Version of the service being requested (will be <= the version advertised)</param>
        /// <param name="aWidgetUdn"></param>
        /// <param name="aWidgetClass"></param>
        protected virtual void CreateWidget(uint aVersion, string aWidgetUdn, uint aWidgetClass)
        {
            throw (new ActionDisabledError());
        }

        /// <summary>
        /// RemoveWidget action.
        /// </summary>
        /// <remarks>Will be called when the device stack receives an invocation of the
        /// RemoveWidget action for the owning device.
        ///
        /// Must be implemented iff EnableActionRemoveWidget was called.</remarks>
        /// <param name="aVersion">Version of the service being requested (will be <= the version advertised)</param>
        /// <param name="aWidgetUdn"></param>
        protected virtual void RemoveWidget(uint aVersion, string aWidgetUdn)
        {
            throw (new ActionDisabledError());
        }

        /// <summary>
        /// SetWidgetRegister action.
        /// </summary>
        /// <remarks>Will be called when the device stack receives an invocation of the
        /// SetWidgetRegister action for the owning device.
        ///
        /// Must be implemented iff EnableActionSetWidgetRegister was called.</remarks>
        /// <param name="aVersion">Version of the service being requested (will be <= the version advertised)</param>
        /// <param name="aWidgetUdn"></param>
        /// <param name="aRegisterIndex"></param>
        /// <param name="aRegisterValue"></param>
        protected virtual void SetWidgetRegister(uint aVersion, string aWidgetUdn, uint aRegisterIndex, uint aRegisterValue)
        {
            throw (new ActionDisabledError());
        }

        /// <summary>
        /// GetWidgetRegister action.
        /// </summary>
        /// <remarks>Will be called when the device stack receives an invocation of the
        /// GetWidgetRegister action for the owning device.
        ///
        /// Must be implemented iff EnableActionGetWidgetRegister was called.</remarks>
        /// <param name="aVersion">Version of the service being requested (will be <= the version advertised)</param>
        /// <param name="aWidgetUdn"></param>
        /// <param name="aRegisterIndex"></param>
        /// <param name="aRegisterValue"></param>
        protected virtual void GetWidgetRegister(uint aVersion, string aWidgetUdn, uint aRegisterIndex, out uint aRegisterValue)
        {
            throw (new ActionDisabledError());
        }

        private static unsafe int DoCreateWidget(IntPtr aPtr, IntPtr aInvocation, uint aVersion)
        {
            GCHandle gch = GCHandle.FromIntPtr(aPtr);
            DvProviderZappOrgTestWidgetController1 self = (DvProviderZappOrgTestWidgetController1)gch.Target;
            DvInvocation invocation = new DvInvocation(aInvocation);
            string widgetUdn = invocation.ReadString("WidgetUdn");
            uint widgetClass = invocation.ReadUint("WidgetClass");
            self.CreateWidget(aVersion, widgetUdn, widgetClass);
            invocation.WriteStart();
            invocation.WriteEnd();
            return 0;
        }

        private static unsafe int DoRemoveWidget(IntPtr aPtr, IntPtr aInvocation, uint aVersion)
        {
            GCHandle gch = GCHandle.FromIntPtr(aPtr);
            DvProviderZappOrgTestWidgetController1 self = (DvProviderZappOrgTestWidgetController1)gch.Target;
            DvInvocation invocation = new DvInvocation(aInvocation);
            string widgetUdn = invocation.ReadString("WidgetUdn");
            self.RemoveWidget(aVersion, widgetUdn);
            invocation.WriteStart();
            invocation.WriteEnd();
            return 0;
        }

        private static unsafe int DoSetWidgetRegister(IntPtr aPtr, IntPtr aInvocation, uint aVersion)
        {
            GCHandle gch = GCHandle.FromIntPtr(aPtr);
            DvProviderZappOrgTestWidgetController1 self = (DvProviderZappOrgTestWidgetController1)gch.Target;
            DvInvocation invocation = new DvInvocation(aInvocation);
            string widgetUdn = invocation.ReadString("WidgetUdn");
            uint registerIndex = invocation.ReadUint("RegisterIndex");
            uint registerValue = invocation.ReadUint("RegisterValue");
            self.SetWidgetRegister(aVersion, widgetUdn, registerIndex, registerValue);
            invocation.WriteStart();
            invocation.WriteEnd();
            return 0;
        }

        private static unsafe int DoGetWidgetRegister(IntPtr aPtr, IntPtr aInvocation, uint aVersion)
        {
            GCHandle gch = GCHandle.FromIntPtr(aPtr);
            DvProviderZappOrgTestWidgetController1 self = (DvProviderZappOrgTestWidgetController1)gch.Target;
            DvInvocation invocation = new DvInvocation(aInvocation);
            string widgetUdn = invocation.ReadString("WidgetUdn");
            uint registerIndex = invocation.ReadUint("RegisterIndex");
            uint registerValue;
            self.GetWidgetRegister(aVersion, widgetUdn, registerIndex, out registerValue);
            invocation.WriteStart();
            invocation.WriteUint("RegisterValue", registerValue);
            invocation.WriteEnd();
            return 0;
        }

        /// <summary>
        /// Must be called for each class instance.  Must be called before Core.Library.Close().
        /// </summary>
        public void Dispose()
        {
            DoDispose();
            GC.SuppressFinalize(this);
        }

        ~DvProviderZappOrgTestWidgetController1()
        {
            DoDispose();
        }

        private void DoDispose()
        {
            lock (this)
            {
                if (iHandle == IntPtr.Zero)
                {
                    return;
                }
                DisposeProvider();
                iHandle = IntPtr.Zero;
            }
            iGch.Free();
        }
    }
}

